### PR TITLE
android [nfc]: Move notif intent-handling function into notifications/

### PR DESCRIPTION
**Pull request by chrisbobbe**
Toward https://github.com/zulip/zulip-mobile/issues/5693, after https://github.com/zulip/zulip-mobile/pull/5715.

### CODE REVIEW

File: android/app/src/main/java/com/zulipmobile/MainActivity.kt [hunk 1]
```diff
@@ -30,26 +29,23 @@ open class MainActivity : ReactActivity() {
...
             }
 
-            Intent.ACTION_VIEW -> when {
-                // Launch MainActivity on tapping a notification
-                url?.scheme == "zulip" && url.authority == NOTIFICATION_URL_AUTHORITY -> {
-                    val data = intent.getBundleExtra(EXTRA_NOTIFICATION_DATA) ?: return false
-                    logNotificationData("notif opened", data)
-                    notifyReact(reactApplication.tryGetReactContext(), data)
+            Intent.ACTION_VIEW -> {
+                if (maybeHandleViewIntent(intent, maybeReactContext)) {
```
Comment (chrisbobbe) [inline android/app/src/main/java/com/zulipmobile/MainActivity.kt:41]: Does the function `maybeHandleViewIntent` want a name that mentions notifications, so it's clear what it means in the context of surrounding code, which isn't all about notifications?

File: android/app/src/main/java/com/zulipmobile/MainActivity.kt [hunk 2]
```diff
@@ -30,26 +29,23 @@ open class MainActivity : ReactActivity() {
     /* Returns true just if we did handle the intent. */
     private fun maybeHandleIntent(intent: Intent?): Boolean {
         intent ?: return false
-        val url = intent.data
+        val maybeReactContext = reactApplication.tryGetReactContext()
         when (intent.action) {
             // Share-to-Zulip
             Intent.ACTION_SEND, Intent.ACTION_SEND_MULTIPLE -> {
-                handleSend(intent, reactApplication.tryGetReactContext(), contentResolver)
+                handleSend(intent, maybeReactContext, contentResolver)
```
Comment (gnprice) [inline android/app/src/main/java/com/zulipmobile/MainActivity.kt:36]: I don't think this benefits from getting the React context eagerly; we may as well wait to invoke `reactApplication.tryGetReactContext()` until we know we're going down a codepath that needs it. (IOW, can drop the "Pull out a `maybeReactContext` variable" commit.)

File: android/app/src/main/java/com/zulipmobile/MainActivity.kt (hunk 1 above)
Comment (gnprice) [reply android/app/src/main/java/com/zulipmobile/MainActivity.kt:41]: Yeah. How about `maybeHandleViewNotif`? The particular "view" intents that it does handle are those that are about a notification — those where the URL identifying what the user wanted to view is a URL in a form we made up to mean "open a notification".

File: android/app/src/main/java/com/zulipmobile/notifications/NotifyReact.kt [hunk 3]
```diff
@@ -17,6 +16,26 @@ import com.zulipmobile.emitEvent
  * TODO: Replace this with a fresh implementation based on RN upstream docs.
  */
 
+/**
+ * Recognize if an ACTION_VIEW intent came from tapping a notification; handle it if so
+ *
+ * Sends a 'notificationOpened' event to the JavaScript layer.
+ */
+internal fun maybeHandleViewIntent(intent: Intent, maybeReactContext: ReactContext?): Boolean {
```
Comment (gnprice) [inline android/app/src/main/java/com/zulipmobile/notifications/NotifyReact.kt:24]: A key piece of this function's interface which the doc comment should mention is that it returns true just if it did handle the intent. (Just like the comment on `maybeHandleIntent`.)

**Review by gnprice — COMMENTED** [review]
Thanks! Small comments.

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed.

**Comment by gnprice** [comment]
Thanks! Looks good; merging.

**Pull request merged**